### PR TITLE
Support DuckDB style struct syntax

### LINKED ISSUE #9820: Support DuckDB style stuct syntax
### Is your feature request related to a problem or challenge?

To make it easier to use structs with DataFusion, having a nicer literal syntax is useful

### Describe the solution you'd like

It would be nice to support duckdb style struct syntax:

https://duckdb.org/docs/sql/data_types/struct

```sql
SELECT {'yes': 'duck', 'maybe': 'goose', 'huh': NULL, 'no': 'heron'};
```

### Describe alternatives you've considered

One way to support this feature would be to implement some `Expr` variant for the struct literal,

### Additional context

@gstein  is adding `named_struct` in this function https://github.com/apache/arrow-datafusion/pull/9743

@jayzhan211  filed https://github.com/sqlparser-rs/sqlparser-rs/pull/1194 for the upstream sqlparser support

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #9820 .

### CODE REVIEW

File: datafusion/sql/src/expr/mod.rs [hunk 1]
```diff
@@ -619,10 +620,45 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        for planner in self.planners.iter() {
+            match planner.plan_dictionary_literal(raw_expr, schema)? {
+                PlannerResult::Planned(expr) => {
+                    return Ok(expr);
+                }
+                PlannerResult::Original(expr) => raw_expr = expr,
+            }
+        }
+
+        internal_err!("Expected a simplified result, but none was found")
```
Comment (jonahgao) [inline datafusion/sql/src/expr/mod.rs:659]: Would it be better to return a `not_impl_err` error?
User may only use `SqlToRel` without a `SessionState`.

**Review by alamb — APPROVED** [review]
Thanks @jayzhan211  -- this is so cool!

```sql
DataFusion CLI v39.0.0
> select {'foo': 1};
+------------------------------------+
| named_struct(Utf8("foo"),Int64(1)) |
+------------------------------------+
| {foo: 1}                           |
+------------------------------------+
1 row(s) fetched.
Elapsed 0.015 seconds.

> select {'foo': {'bar': 2}};
+--------------------------------------------------------------+
| named_struct(Utf8("foo"),named_struct(Utf8("bar"),Int64(2))) |
+--------------------------------------------------------------+
| {foo: {bar: 2}}                                              |
+--------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.002 seconds.
```

I took the liberty of merging up from main to resolve a conflict, and since I had the PR open I also implemented @jonahgao 's suggestion and updated some comments. Hope you don't mind

THank you again

File: datafusion/core/src/execution/session_state.rs [hunk 2]
```diff
@@ -231,26 +231,16 @@ impl SessionState {
...
-            user_defined_sql_planners.extend(vec![array_planner, field_access_planner]);
-        }
-        #[cfg(feature = "datetime_expressions")]
-        {
-            let extract_planner =
-                Arc::new(functions::datetime::planner::ExtractPlanner::default()) as _;
-
-            user_defined_sql_planners.push(extract_planner);
-        }
+        let user_defined_sql_planners: Vec<Arc<dyn UserDefinedSQLPlanner>> = vec![
```
Comment (alamb) [inline datafusion/core/src/execution/session_state.rs:234]: I also took the opportunity while I had this PR checked out to resolve a conflict to consolidate the code in the constructor

cc @dharanad  and @xinlifoobar

**Comment by jayzhan211** [comment]
Great! 💯 
Thanks @alamb and @jonahgao

**Pull request merged**